process/smartContract/process.go

```diff
@@ -910,11 +903,37 @@ func (sc *scProcessor) createSCRsWhenError(
 		Value:         tx.GetValue(),
 		RcvAddr:       rcvAddress,
 		SndAddr:       tx.GetRcvAddr(),
-		Data:          []byte("@" + hex.EncodeToString([]byte(returnCode)) + "@" + hex.EncodeToString(txHash)),
 		PrevTxHash:    txHash,
 		ReturnMessage: returnMessage,
 	}
 
+	consumedFee := core.SafeMul(tx.GetGasLimit(), tx.GetGasPrice())
+	if !sc.flagDeploy.IsSet() {
+		scr.Data = []byte("@" + hex.EncodeToString([]byte(returnCode)) + "@" + hex.EncodeToString(txHash))
+		if check.IfNil(acntSnd) {
+			moveBalanceCost := sc.economicsFee.ComputeMoveBalanceFee(tx)
+			consumedFee.Sub(consumedFee, moveBalanceCost)
+		}
+	} else {
+		if callType == vmcommon.AsynchronousCall {
+			scr.CallType = vmcommon.AsynchronousCallBack
+			scr.GasPrice = tx.GetGasPrice()
+
+			gasToLock := sc.asyncCallStepCost + sc.asyncCallbackGasLock
+			if tx.GetGasLimit() >= gasToLock {
```

Comment:
all the gasLimit is consumed here. in the cross shard it will have only the tx.Value added back to the original sender, and that's it.

Reply:
I mean gasToLock > tx.GetGasLimit()

Reply:
It could be the situation that asyncCallback to not be executed because the tx.GetGasLimit() < gasToLock, and in this way the tx value will not be returned?

Reply:
In that case the asyncCallback reaches the initial sender shard and value is returned. Even in case of error an asyncCallBack will add the txValue to the sender.

Reply:
ok

Reply:
Normally, `tx.GetGasLimit() < gasToLock` cannot happen, because Arwen performs this check before initiating the `asyncCall()`. It should still be verified here, though.